[MRG] Release 0.20.1

jnothman opened the pull request:
This is gathering commits for 0.20.1. See https://github.com/scikit-learn/scikit-learn/milestone/27 for remaining work. From master == f5ef674f3d02df2b4921052a5e97e559014f1ce2, I have performed the following interactive rebase:

(Devs take note: it's really nice if the first line of a commit message makes it clear what component it relates to, as well as whether it's DOC/FIX/TST, etc. Feel free to rename PRs or edit the commit message upon squash-and-merge, although making sure not to remove the PR number.)

```
drop 27972d4 MNT bump to version 0.21.dev0 after branching 0.20.X (#11941)
drop 5a10991 CI Handle new branch in scikit-learn.github.io (#11945)
drop a47e779 EXA Fix plot in plot_optics.py
drop 8cbb131 MNT joblib 0.12.3 (#11949)
drop 0908617 typo and formatting fixes in 0.20 doc (#11963)
drop f575384 DOC whats new boilerplate for 0.21 (#11974)
drop 721ebae MNT Change max_bound -> max_eps in OPTICS (#11984)
drop 84c4e54 COSMIT remove unnecessary _TreeNode methods (#11983)
drop 07051bc DOC OPTICS: improve docstring and add default values. (#11987)
drop c268230 MNT Remove n_clusters_ in OPTICS (#11981)
drop 5e101a2 Joblib 0.12.4 (#12007)
drop e5333f5 OPTICS remove redundant recursion (#11985)
drop 251e58b FIX ordering_ type and cosmetic changes to structure for OPTICS main loop (#11986)
drop a86709f [MRG] MNT rename min_cluster_size_ratio to min_cluster_size (#11913)
drop a79d44e FIX OPTICS Change quick_scan floating point comparison to isclose (#11929)
drop efe7b8c DOC Fix optics metric issues (DOC and precomputed) (#12028)
drop e616ee3 DOC move OPTICS to 0.21
drop daaa2a5 DOC Reword to avoid that people draw wrong conclusions (#12095)
drop 5d30b2d ENH (0.21) Make OPTICS more memory efficient when calling kneighbors (#12103)
drop e81bcd5 Added the changes to remove the documentation support statements for Python 2 (#12083)
drop 755c7dd DOC (0.21) OPTICS Note the order of reachability_ and core_distances_ (#12132)
pick 09851ac TST update make_column_transformer test + add comment (#12156)
pick 9427c36 coef0 is a float, not an int (#12161)
pick e63feeb DOC More specific about the limitation of make_column_transformer (#12163)
pick b886da5 MAINT update comment
pick b915ca6 MNT Avoid using "is" when comparing strings (#12168)
pick 6463406 MNT Unused import in plot_gpr_co2.py
drop 4207225 DOC start section for the 0.20.1 bugfix notes (#12170)
pick 88b49e5 Fix parallel backend neighbors (#12172)
drop a11154e [MRG] improve check_non_negative for sparse matrices (#12106)
pick da0cb32 FIX Use take instead of choose in compute_sample_weight (#12165)
pick d88bef1 DOC Add sections to whats new 0.20.1 (#12183)
pick 2e2e69d DOC KDE normalisation clarified (#11275)
pick 819d8ef [MRG] Fix diagonal in DBSCAN with precomputed sparse neighbors graph (#12105)
squash ea52161 MNT Move what's new entry
pick a358d7d DOC Add Versionadded tag to sklearn/_config.py (#12187)
pick 239482f BaseSearchCV._run_search raises NotImplementedError instead of being an abstractmethod (#12182)
pick 2cf145d DOC Add versionadded to set_config (#12196)
pick 0b58bc3 DOC Improve ColumnTransformer docstrings (#12206)
pick 9d58ca5 MNT Remove duplicate import of warnings & unused variables (#12203)
pick 663e024 DOC Fix typo in neighbors/nearest_centroid.py (#12223)
pick e1c3c22 DOC Fixing summary table in the linear model documentation. (#12220)
pick a3616f6 MNT Use name instead of float to specify colors (#12199)
pick 94c70ff [MRG] More informative error message in OneHotEncoder(categories=None) with negative integer values (#12180)
pick 59b15c5 add explicit mention of scaing for saga in logisticregression docs. (#12236)
pick 11612fc MNT Raise error for duplicate classes when constructing a MultiLabelBinarizer (#12195)
pick 5313325 DOC Encourage contributors to use sklearn.show_versions() (#12225)
pick acf3bab MNT Add versionadded to set_config parameters
pick dd3b705 MNT Unused imports in examples
pick 60cf1d6 Fix numpy.int overflow in make_classification (#10811)
squash da85815 DOC what's new entry for "Fix numpy.int overflow in make_classification #10811"
squash 877e3f3 MNT Move what's new entry
drop 7166cd5 MNT Remove duplicate entry in whats new
pick dfd009d Remove test_import_sklearn_no_warnings (#12244)
pick cbbe489 DOC fix cross-entropy typo in tree docs (#12242)
pick bfab306 [MRG] Added Tips in SVM user guide for tuning C parameter in LinearSVC and LinearSVR (#12185)
drop 3e5777a [MRG] Fast PolynomialFeatures on dense arrays (#12251)
drop fb7be87 [MRG + 1] return_train_score deprecation (#12241)
drop e0e7387 Remove unused private functions (#12253)
pick 24aa6b8 DOC Remove mentions of removed AUTHORS.rst file (#12262)
pick f456a40 MNT Change show_versions format to suit markdown (#12255)
pick 5e24762 DOC add note on discretization creating non-linearity (#12269)
pick e9cdb55 MNT Updated PyPI URLs (#12274)
drop 2eca77b MNT complete VotingClassifier flatten_transform deprecation (#12252)
pick 8e08028 DOC Move 'for instance' to front
pick 58228cb [MRG+1] ColumnTransformer fix having mixed types in a single passthrough (#12200)
pick b725921 [MRG] Created 'cross-validation estimator' entry in glossary (#11661)
pick 3f5bf97 DOC removed ambiguity in pipeline gridsearch example (#12272)
pick 74b56db MNT Converting http to https (#12277)
pick 1e052e9 Converting http to https (2)... (#12292)
pick 0bbb7d0 DOC Add class example for LedoitWolf (#12214)
pick afa0694 FIX cache of OpenML fetcher (#12246)
pick 08924c3 EXA Fix bad data visualisation in "Importance of Feature Scaling" (#12280)
pick a1be325 MNT Make check_X_y raise a better error when y is None (#12283)
pick 8afe43e DOC Added "mars" testimonial to testimonials page (#12298)
pick 9e1d48f DOC What's new typo
pick e6359e2 DOC Fix broken link to joblib documentation (#12301)
pick b020f62 MNT Apply pep8 to docs code (#12275)
drop f4e7d2b Converting http to https (3)... (#12302)
pick 3804ccd MNT Refactors doc test into seperate script (#12248)
drop 4280308 ENH (0.21) Remember predecessor in OPTICS (#12135)
drop 2020867 FIX (0.21) OPTICS correctly handle multiple infs in reachability array. (#12029)
pick 63e5ae6 TST Use same random seeds for both GaussianMixture.fit (#12307)
drop 4e2e1fa ENH Cache class mapping in MultiLabelBinarizer() (#12116)
pick 5fd9e03 ENH Raise an error when pos_label is not in binary y_true (#12313)
pick c5b020f TST Use v_measure_score to compare label arrays up to permutation (#12265)
pick 205ff38 DOC What's new typo
pick 5df8cd3 EXA Fix title, overlapping plots and axis labels in plot_ols_ridge_variance.py (#12296)
pick 03c3af5 [MRG] Fix fetch_openml when ignore attributes are numeric (#12330)
pick bbb0d93 [MRG] FIX Update power_transform docstring and add FutureWarning (#12317)
pick c8a4132 DOC check_array() and check_X_y() documentation update (#12340)
drop a80bbd9 ENH add get_n_leaves() and get_max_depth() to DesicionTrees (#12300)
pick 00c2f41 DOC fix logistic regression.fit docstring on y (#12343)
drop 39bd736 [MRG] Move RandomTreesEmbedding criterion & max_features to be class attributes (#12324)
drop 831c760 ENH (0.21) Add max_error to the existing set of metrics for regression (#12232)
pick 8985a63 DOC Update v0.20.rst with power_transform API change (#12351)
drop 0f94f29 MNT simple deprecations and removals for 0.21 (#12238)
drop c13ba26 [MRG] Matplotlib tree plotting (#9251)
pick c76dc05 EXA Fix labels overlap in Out-of-core classification of text documents example (#12359)
pick 1e7cd7d BUG Fix OrdinalEncoder with manually specified categories (#12367)
pick 76b1078 TST Fix missing assert and parametrize k-means tests  (#12368)
drop 0296916 FIX (0.21) OPTICS processing order (#12357)
pick 0ad8736 [MRG] ENH apply sparse_threshold even if all columns are sparse (#12304)
drop 98357ec FIX (0.21) make count_nonzero dtype invariant wrt axis (#12341)
drop 274ce3a DOC use default_role='any' (#12355)
pick f5ef674 DOC improve OneHotEncoder documentation (#12314)
pick 906f078 DOC Fix the link to computing docs (#12380)
drop afe0a9b MNT skip test falling on master in legacy platforms (#12382)
pick a1d0e96 FIX Increase mean precision for large float32 arrays (#12338)
pick bea1eb5 [MRG] Clarified indempotence of fit (#12305)
pick d4c9e84 DOC minor clarifications in ensemble.rst (#11810)
pick d9b56b6 DOC DecisionTreeClassifier does not support categorical data (#12402)
pick 53069c2 DOC Change i.e. to e.g. in MinMaxScaler (#12415)
drop a5fa7d3 [MRG] Fast PolynomialFeatures on CSR matrices (#12197)
drop 5bcd84b [MRG+1] Add check_is_fitted to non standard functions (#12279)
pick 99bea79 DOC: Fix typo where FMI was referred to as AMI (#12414)
drop 3c76b9c FIX 'MultiTaskLasso' object has no attribute 'coef_' when warm_start = True (#12361)
drop 5af272a MNT Remove unused variables (#12230)
pick 2ab1961 Resurrected PR #5224 from @andylamb. (#12427)
pick 81d2178 EXA calculate number of noise points (#12428)
pick b498ac7 ENH Raise descriptive ValueError if number of samples equals number of classes in Linear Discriminant Analysis (#12391)
drop be742c3 MNT Change default metric in OPTICS (#12439)
pick 013d295 FIX olivetti_faces DESCR to point to the good location (#12441)
drop 88cdeb8 FIX (0.21) Correct RD/CD/order/predecessor in OPTICS (#12421)
drop dc883c7 ENH Corrected spelling of Harabasz score. (#12211)
pick ebe77d6 modify kbins test using kmeans due to unstable local minimum (#12450)
pick 4e2da4a [MRG] Added FutureWarning in sgd models for tol parameter (#12399)
pick e50fc2a TST Fix test gaussian mixture warm start (#12452)
drop 9ec5a15 ALL Add resample and shuffle to __all__ (#12456)
pick 5cef1df FIX ensure max_features > 0 in ensemble.bagging (#12388)
pick 2912e3a TST Parametrize, refactor and add new kmeans tests (#12432)
pick 338f763 TST Throw correct error for pytest version (#12475)
pick e67e30c Fix numpy vstack on generator expressions (#12467)
pick d4802ae Fix mean shift equation as per issue 12420 (#12455)
drop f6b0c67 TST Added estimator check for idempotence of fit() (#12328)
pick a1fabce DOC Mention of pairwise_distances in Guide on Metrics (#12416)
drop 6555631 FEA (0.21) multilabel confusion matrix (#11179)
pick c2f17d0 FIX pairwise_distances_argmin_min wrong with metric="euclidean" (#12481)
pick 478267e DOC Small fix in compose.rst (#12487)
pick 83ac5dc MNT what's new corrections
pick c6455fa Fix include .pxd definitions for sklearn.tree in wheels (#12381)
pick 1c88b3c Fix IncrementalPCA when final batch is smaller than dimensions required for SVD (#12379)
pick d42d0d7 DOC fix broken link in doc/developers/contributing.rst (#12508)
pick 6226cf5 TST skip test_backend_respected if joblib is forced into serial mode (#12496)
pick 036dfdd Fix inconsistent labels returned by BayesianGaussianMixture.fit_predict (#12451)
pick a028416 DOC Add details to StandardScaler calculation (#12446)
drop 9f842e3 API Adds passthrough option to Pipeline (#11674)
pick f24e300 Impose shared memory when fitting a SGDClassifier (#12498)
edit 6d9acd7 MNT what's new corrections
pick 6b4e00d MNT KBinsDiscretizer.transform should not mutate _encoder (#12514)
pick 5d8dfc9 FIX SkLearn `.score()` method generating error with Dask DataFrames (#12462)
drop c676981 MNT (0.21) OPTiCS change the default `algorithm` to `auto` (#12529)
pick 5196657 [MRG] Fix segfault in AgglomerativeClustering with read-only mmaps (#12485)
pick 6f5fec9 Fix dead link to numpydoc (#12532)
drop 8e3a191 fix typo in whatsnew
pick 1128094 BLD we should ensure continued support for joblib 0.11 (#12350)
drop 8408b46 ALL Add HashingVectorizer to __all__ (#12534)
drop 54de189 DOC (0.21) Make sure plot_tree docs are generated and fix link in whatsnew (#12533)
pick c68f306 DOC tweak KMeans regarding cluster_centers_ convergence (#12537)
pick c81e255 joblib 0.13.0 (#12531)
drop 664ff34 ENH Prefer threads for IsolationForest (#12543)
pick 3282d43 [MRG] Additional Warnings in case OpenML auto-detected a problem with dataset  (#12541)
pick 042843a FIX YeoJohnson transform lambda bounds (#12522)
pick ea115c2 DOC: add a testimonial from JP Morgan (#12555)
pick 362cb3b TST autoreplace assert_true(...==...) with plain assert (#12547)
pick 43e3a02 MNT Remove unused assert_true imports (#12560)
pick eb36e49 MNT Don't change self.n_values in OneHotEncoder.fit (#12286)
pick 8622885 DOC Add skorch to related projects (#12561)
pick eb36c28 FIX stop words validation in text vectorizers with custom preprocessors / tokenizers (#12393)
pick 6c5f285 EXA Fix comment in plot-iris-logistic example (#12564)
pick 7c2b47a DOC Add 's' to "correspond" in docs for Hamming Loss. (#12565)
pick 4e81949 FIX remove FutureWarning in _object_dtype_isnan and add test (#12567)
pick 4b78d7a DOC: Clarify `cv` parameter description in `GridSearchCV` (#12495)
pick 2afee93 FIX incorrect error when OneHotEncoder.transform called prior to fit (#12443)
pick 1f2dd75 MNT bare asserts (#12571)
pick 32e5fd4 FIX Workaround limitation of cloudpickle under PyPy (#12566)
pick 914a915 DOC Fix typo (#12563)
pick d0e99db TST don't test utils.fixes docstrings (#12576)
drop 02dc9ed Fix max_depth overshoot in BFS expansion of trees (#12344)
drop 4b0a70c TST Use NotFittedError instead of ValueError in test_dummy.py (#12579)
pick 14c816e ENH/FIX openml, Adds retrying if reading from cache fails (#12526)
pick 94db3d9 ENH Improved error message for bad predict_proba shape in ThresholdScorer (#12486)
pick bfc4a56 MNT Duplicate import
pick c47c8a9 DOC improved documentation of MissingIndicator (#12424)
drop 24e4641 FIX: clone behavior for estimator types (#12585)
pick 0b85b0a FIX use ellipsis in PowerTransformer doctest (#12595)
pick feed6c7 DOC fix typos in gaussian_process.rst (#12602)
pick 1d57955 DOV consistency of parameters for GroupKFold and LeaveOneGroupOut (#12581)
pick 82e095c DOC generative model description etc for LatentDirichletAllocation (#12216)
pick 62117f4 Bug with string dtype in OneHotEncoder with handle_unknown='ignore' (#12471)
pick 66507a2 MAINT Use CircleCI for linting (#12606)
pick 9ecada8 DOC fix TeX error in Davies Bouldin formula
pick 4ab6055 MNT Change deprecation for min_impurity_split from removal to changing the default (#12400)
pick d25da1b FIX make joblib utils private, and remove mentions of externals.joblib (#12345)
pick ac327c5 [MRG] DOC Windows build dependencies (#12615)
pick 0b8650a [MRG] set blockwise diagonals to zero for euclidean distance (#12612)
pick 104f684 FIX check_array dtype check for pandas series (#12625)
```

jorisvandenbossche committed:
TST update make_column_transformer test + add comment (#12156)

Follow-up on https://github.com/scikit-learn/scikit-learn/pull/12152
And added comment why transformer_weights is not passed through, see
https://github.com/scikit-learn/scikit-learn/pull/11183#pullrequestreview-125539051
for more discussion

djsutherland committed:
coef0 is a float, not an int (#12161)

qinhanmin2014 committed:
DOC More specific about the limitation of make_column_transformer (#12163)

jnothman committed:
MAINT update comment

qinhanmin2014 committed:
MNT Avoid using "is" when comparing strings (#12168)

qinhanmin2014 committed:
MNT Unused import in plot_gpr_co2.py

tomMoral committed:
Fix parallel backend neighbors (#12172)

jnothman committed:
FIX Use take instead of choose in compute_sample_weight (#12165)

qinhanmin2014 committed:
DOC Add sections to whats new 0.20.1 (#12183)

haroldfox committed:
DOC KDE normalisation clarified (#11275)

TomDLT committed:
[MRG] Fix diagonal in DBSCAN with precomputed sparse neighbors graph (#12105)

J0 committed:
DOC Add Versionadded tag to sklearn/_config.py (#12187)

adrinjalali committed:
BaseSearchCV._run_search raises NotImplementedError instead of being an abstractmethod (#12182)

* _run_search raises NotImplementedError instead of being and abstractmethod

* add error message

* test for a BaseSearchCV child w/o a _run_search

* make the test python2 compatible, still in 0.20 zone.

* specify cv in tests not to trigger the related FutureWarning

* PEP8

qinhanmin2014 committed:
DOC Add versionadded to set_config (#12196)

LilyX2021 committed:
DOC Improve ColumnTransformer docstrings (#12206)

rebekahkim committed:
MNT Remove duplicate import of warnings & unused variables (#12203)

Commit:
DOC Fix typo in neighbors/nearest_centroid.py (#12223)

markhannel committed:
DOC Fixing summary table in the linear model documentation. (#12220)

qinhanmin2014 committed:
MNT Use name instead of float to specify colors (#12199)

ygivenx committed:
[MRG] More informative error message in OneHotEncoder(categories=None) with negative integer values (#12180)

* Fix Issue #12179

OneHotEncoder "only non-negative integers" message should suggest using
categories='auto'

* Fix Issue #12179

    OneHotEncoder "only non-negative integers" message should suggest using
    categories='auto'

* Fix Issue #12179

OneHotEncoder "only non-negative integers" message should suggest using
categories='auto'

* Fixes #12180 Modify the error message

* Fix the spacing

amueller committed:
add explicit mention of scaing for saga in logisticregression docs. (#12236)

samwaterbury committed:
MNT Raise error for duplicate classes when constructing a MultiLabelBinarizer (#12195)

qinhanmin2014 committed:
DOC Encourage contributors to use sklearn.show_versions() (#12225)

qinhanmin2014 committed:
MNT Add versionadded to set_config parameters

qinhanmin2014 committed:
MNT Unused imports in examples

VarIr committed:
Fix numpy.int overflow in make_classification (#10811)

rth committed:
Remove test_import_sklearn_no_warnings (#12244)

rasbt committed:
DOC fix cross-entropy typo in tree docs (#12242)

Commit:
[MRG] Added Tips in SVM user guide for tuning C parameter in LinearSVC and LinearSVR (#12185)

* Added Tip for tuning C parameter in LinearSVC/SVR

* Added reference about performance improvement

* Clearer explanation about large C values

* Update svm.rst

jacquerie committed:
DOC Remove mentions of removed AUTHORS.rst file (#12262)

qinhanmin2014 committed:
MNT Change show_versions format to suit markdown (#12255)

jnothman committed:
DOC add note on discretization creating non-linearity (#12269)

TakingItCasual committed:
MNT Updated PyPI URLs (#12274)

jnothman committed:
DOC Move 'for instance' to front

adrinjalali committed:
[MRG+1] ColumnTransformer fix having mixed types in a single passthrough (#12200)

NicolasHug committed:
[MRG] Created 'cross-validation estimator' entry in glossary (#11661)

* Created 'cross-validation estimator' in glossary and referenced it where
needed

* Addressed adrinjalali comments

* updated glossary entry

* updaed docstrings according to commets

sjtrny committed:
DOC removed ambiguity in pipeline gridsearch example (#12272)

* Removed ambiguity in pipeline gridsearch example

* Formatted to pep8

TakingItCasual committed:
MNT Converting http to https (#12277)

TakingItCasual committed:
Converting http to https (2)... (#12292)

andreanr committed:
DOC Add class example for LedoitWolf (#12214)

janvanrijn committed:
FIX cache of OpenML fetcher (#12246)

benjastudio committed:
EXA Fix bad data visualisation in "Importance of Feature Scaling" (#12280)

jacquerie committed:
MNT Make check_X_y raise a better error when y is None (#12283)

gauravahlawat01goal committed:
DOC Added "mars" testimonial to testimonials page (#12298)

qinhanmin2014 committed:
DOC What's new typo

kushalchauhan98 committed:
DOC Fix broken link to joblib documentation (#12301)

kataev committed:
MNT Apply pep8 to docs code (#12275)

thomasjpfan committed:
MNT Refactors doc test into seperate script (#12248)

oleksandr-pavlyk committed:
TST Use same random seeds for both GaussianMixture.fit (#12307)

qinhanmin2014 committed:
ENH Raise an error when pos_label is not in binary y_true (#12313)

oleksandr-pavlyk committed:
TST Use v_measure_score to compare label arrays up to permutation (#12265)

qinhanmin2014 committed:
DOC What's new typo

Commit:
EXA Fix title, overlapping plots and axis labels in plot_ols_ridge_variance.py (#12296)

janvanrijn committed:
[MRG] Fix fetch_openml when ignore attributes are numeric (#12330)

* modularized data column functionality

* small bugfix

* removes redundant line breaks

* added some documentation on the added fn

* added additional comment on advice of Nicholas Hug

* added test case

* merged master into branch, and added small comments by Joel

* added doc item

chang committed:
[MRG] FIX Update power_transform docstring and add FutureWarning (#12317)

dvukolov committed:
DOC check_array() and check_X_y() documentation update (#12340)

adrinjalali committed:
DOC fix logistic regression.fit docstring on y (#12343)

chang committed:
DOC Update v0.20.rst with power_transform API change (#12351)

Commit:
EXA Fix labels overlap in Out-of-core classification of text documents example (#12359)

jorisvandenbossche committed:
BUG Fix OrdinalEncoder with manually specified categories (#12367)

jeremiedbb committed:
TST Fix missing assert and parametrize k-means tests  (#12368)

amueller committed:
[MRG] ENH apply sparse_threshold even if all columns are sparse (#12304)

Cheukting committed:
DOC improve OneHotEncoder documentation (#12314)

gustavodemari committed:
DOC Fix the link to computing docs (#12380)

jnothman committed:
MNT Update version

bauks committed:
FIX Increase mean precision for large float32 arrays (#12338)

amueller commented:
Is there a reason you rebase instead of cherry-pick?

NicolasHug committed:
[MRG] Clarified indempotence of fit (#12305)

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

As far as I understand it, the fact that `fit` is idempotent means that repeated calls to `fit` with the same data doesn't change the estimator.

The contributing guide was a bit unclear about this.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jnothman commented:
> Is there a reason you rebase instead of cherry-pick?

The ultimate effect of rebase and cherry-pick is identical. Interactive rebase does not require you to have commit sha known in advance; it therefore makes it easier to pick commits in bulk, and to be sure that you are thorough. Rebase by default will put the commits in the order they came in.

Interactive rebase also allows you to squash commits together. I've used this above, where a what's new entry had been initially added to 0.21, but then was moved to 0.20.1. Cherry-picking would involve resolving merge conflicts, whereas I think rebase with squash may (at least if there weren't other intervening changes to 0.21) be able recognise that there is no net diff to `v0.21.rst`. Even if I did need to resolve a conflict (by `git rm doc/whats_new/v0.21.rst; git rebase --continue`), this would have been harder and messier to do with cherry-pick.

Finally, rebase -i gives the nice summary above that I can share to document my choices.

jnothman commented:
Also, if I do another rebase of master onto this branch, it is able to recognise which commits I've already transferred to the new branch. So now:

```
$ git checkout master
$ git pull
$ git rebase -i for0.20.1
```

allows me to easily see that there are four commits in master subsequent to the last one I dropped above that are not in this branch. I can drop everything up to 274ce3a and decide whether to drop or pick (or squash) any further commits.

NicolasHug committed:
DOC minor clarifications in ensemble.rst (#11810)

pulkitmaloo committed:
DOC DecisionTreeClassifier does not support categorical data (#12402)

Quetzalcohuatl committed:
DOC Change i.e. to e.g. in MinMaxScaler (#12415)

scottclowe committed:
DOC: Fix typo where FMI was referred to as AMI (#12414)

Commit:
Resurrected PR #5224 from @andylamb. (#12427)

koenvandevelde committed:
EXA calculate number of noise points (#12428)

eddbrown committed:
ENH Raise descriptive ValueError if number of samples equals number of classes in Linear Discriminant Analysis (#12391)

jeremiedbb committed:
FIX olivetti_faces DESCR to point to the good location (#12441)

jeremiedbb committed:
modify kbins test using kmeans due to unstable local minimum (#12450)

NicolasHug committed:
[MRG] Added FutureWarning in sgd models for tol parameter (#12399)

* Added ChangedBehaviorWarning in sgd models

if tol is None while max_iter is set

* Changed to FutureWarning and clarified None meaning

* Ignored warningin tests

* Ignore warnings in tests, round 2

jeremiedbb committed:
TST Fix test gaussian mixture warm start (#12452)

connortann committed:
FIX ensure max_features > 0 in ensemble.bagging (#12388)

jeremiedbb committed:
TST Parametrize, refactor and add new kmeans tests (#12432)

musically-ut committed:
TST Throw correct error for pytest version (#12475)

ogrisel committed:
Fix numpy vstack on generator expressions (#12467)

* Workaround vstack issue with genxp

* Use list comprehensions instead of genexps with np.vstack

* Add changelog entry.

mail-liam committed:
Fix mean shift equation as per issue 12420 (#12455)

#### Reference Issues/PRs
Fixes #12420 

#### What does this implement/fix? Explain your changes.
Updates the mean shift equation as per issue 12420

Commit:
DOC Mention of pairwise_distances in Guide on Metrics (#12416)

jeremiedbb committed:
FIX pairwise_distances_argmin_min wrong with metric="euclidean" (#12481)

thomasjpfan committed:
DOC Small fix in compose.rst (#12487)

qinhanmin2014 committed:
MNT what's new corrections

RoopamSharma committed:
Fix include .pxd definitions for sklearn.tree in wheels (#12381)

minggli committed:
Fix IncrementalPCA when final batch is smaller than dimensions required for SVD (#12379)

CaselIT committed:
DOC fix broken link in doc/developers/contributing.rst (#12508)

yarikoptic committed:
TST skip test_backend_respected if joblib is forced into serial mode (#12496)

ogrisel committed:
Fix inconsistent labels returned by BayesianGaussianMixture.fit_predict (#12451)

tuliocasagrande committed:
DOC Add details to StandardScaler calculation (#12446)

pierreglaser committed:
Impose shared memory when fitting a SGDClassifier (#12498)

TomDLT committed:
MNT what's new corrections

qinhanmin2014 committed:
MNT KBinsDiscretizer.transform should not mutate _encoder (#12514)

Fixes #12490

ZWMiller committed:
FIX SkLearn `.score()` method generating error with Dask DataFrames (#12462)

rth committed:
[MRG] Fix segfault in AgglomerativeClustering with read-only mmaps (#12485)

This fixes a segfault in AgglomerativeClustering with read-only mmaps that happens inside `ward_tree` when calling `scipy.cluster.hierarchy.ward`.

Closes https://github.com/scikit-learn/scikit-learn/issues/12483

(see the above issue for more details)

dstine committed:
Fix dead link to numpydoc (#12532)

jnothman committed:
BLD we should ensure continued support for joblib 0.11 (#12350)

nogilnick committed:
DOC tweak KMeans regarding cluster_centers_ convergence (#12537)

ogrisel committed:
joblib 0.13.0 (#12531)

janvanrijn committed:
[MRG] Additional Warnings in case OpenML auto-detected a problem with dataset  (#12541)

* added additional warning output

* added features gzip

* added gzipped datasets

* fix file naming

* changed expected warning msg

NicolasHug committed:
FIX YeoJohnson transform lambda bounds (#12522)

GaelVaroquaux committed:
DOC: add a testimonial from JP Morgan (#12555)

yarikoptic committed:
TST autoreplace assert_true(...==...) with plain assert (#12547)

qinhanmin2014 committed:
MNT Remove unused assert_true imports (#12560)

amueller committed:
MNT Don't change self.n_values in OneHotEncoder.fit (#12286)

killthekitten committed:
DOC Add skorch to related projects (#12561)

Skorch wraps Pytorch, which makes it awesome. This is a response to [an open issue in Skorch repo](https://github.com/dnouri/skorch/issues/217).

rth committed:
FIX stop words validation in text vectorizers with custom preprocessors / tokenizers (#12393)

lithuak committed:
EXA Fix comment in plot-iris-logistic example (#12564)

JackLangerman committed:
DOC Add 's' to "correspond" in docs for Hamming Loss. (#12565)

jeremiedbb committed:
FIX remove FutureWarning in _object_dtype_isnan and add test (#12567)

QBatista committed:
DOC: Clarify `cv` parameter description in `GridSearchCV` (#12495)

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

This PR addresses issue #12466.

#### What does this implement/fix? Explain your changes.

This PR does the 3 following things:
- Rewrite the `cv` parameter description in `GridSearchCV`
- Link the new `CV splitter` description to an existing example
- Add an example with a custom iterable

Thanks for reviewing this!

Close #12466

dillongardner committed:
FIX incorrect error when OneHotEncoder.transform called prior to fit (#12443)

Commit:
MNT bare asserts (#12571)

ogrisel committed:
FIX Workaround limitation of cloudpickle under PyPy (#12566)

rst0git committed:
DOC Fix typo (#12563)

jnothman committed:
TST don't test utils.fixes docstrings (#12576)

thomasjpfan committed:
ENH/FIX openml, Adds retrying if reading from cache fails (#12526)

Commit:
ENH Improved error message for bad predict_proba shape in ThresholdScorer (#12486)

Continues and resolves #12221, fixes #7598

qinhanmin2014 commented:
@jnothman Travis & Appveyor is failing, seems that you include the tests from https://github.com/scikit-learn/scikit-learn/commit/98357ec4d4e4626cc70e8aaa1380789894d64a93 (and https://github.com/scikit-learn/scikit-learn/commit/afe0a9b70d5ac53836b38b9af3a95659307574fc), but do not include the code change?

qinhanmin2014 committed:
MNT Duplicate import

qinhanmin2014 commented:
I fixed the flake8 error in https://github.com/scikit-learn/scikit-learn/commit/bfc4a566423e036fbdc9fb02765fd893e4860c85

qinhanmin2014 commented:
> @jnothman Travis & Appveyor is failing, seems that you include the tests from 98357ec (and afe0a9b), but do not include the code change?

Maybe it's not harmful to include these two commits, though I'm OK to include them in 0.21.

jnothman commented:
I've changed to drop afe0a9b

jnothman committed:
MNT Duplicate import

amueller commented:
blocker is #12396 mostly, right? There is a weird segfault and even weirder error in ``fetch_openml`` on debian, but no-one has reported those, so I'm not sure if it's worth investigating.
#12345 seems nearly there, so I'd like to include that as well.

janvanrijn committed:
DOC improved documentation of MissingIndicator (#12424)

ogrisel committed:
FIX use ellipsis in PowerTransformer doctest (#12595)

SylvainLan committed:
DOC fix typos in gaussian_process.rst (#12602)

samwaterbury committed:
DOV consistency of parameters for GroupKFold and LeaveOneGroupOut (#12581)

melsyt committed:
DOC generative model description etc for LatentDirichletAllocation (#12216)

GMarzinotto committed:
Bug with string dtype in OneHotEncoder with handle_unknown='ignore' (#12471)

* fix decisionTree issue #12259

* merge splitters

* fix decisionTree issue #12259

* merge splitters

* relaunch tests

* fix for ohe string sizes

* reverse tree changes because not done yet

* error with file

* pylinting

* add regression test and check if categories are strings before casting to object

* pylint

* pylint

* another posibility to handle the strings

* pylint

* pylint

* pylint

* add issue reference and explain the test

* handle object case

* update master test

* update master test

* Update _encoders.py

pylint

* Update test_encoders.py

* Update test_encoders.py

* Update _encoders.py

* Update _encoders.py

* fix checking for object type

* fix checking for object type

* fix checking for object type

* fix checking for object type

* fix encoder

* change check encoder

* add note in whats new

rth committed:
MAINT Use CircleCI for linting (#12606)

jnothman committed:
DOC fix TeX error in Davies Bouldin formula

NicolasHug committed:
MNT Change deprecation for min_impurity_split from removal to changing the default (#12400)

tomMoral committed:
FIX make joblib utils private, and remove mentions of externals.joblib (#12345)

phausamann committed:
[MRG] DOC Windows build dependencies (#12615)

amueller committed:
[MRG] set blockwise diagonals to zero for euclidean distance (#12612)

amueller commented:
why
drop 74b56dbc57 MNT Converting http to https (#12277)
drop 877e3f3235 MNT Move what's new entry
?
Not that it matters much ;)

amueller commented:
#12313 doesn't fix a regression but I think it's a nice bugfix and we should include it.

amueller commented:
Shouldn't #12382 be included?

amueller commented:
and then
```
pick bfc4a56642 MNT Duplicate import
pick c47c8a92e4 DOC improved documentation of MissingIndicator (#12424)
drop 24e46410bd FIX: clone behavior for estimator types (#12585)
pick 0b85b0a922 FIX use ellipsis in PowerTransformer doctest (#12595)
pick feed6c7139 DOC fix typos in gaussian_process.rst (#12602)
pick 1d579556d2 DOV consistency of parameters for GroupKFold and LeaveOneGroupOut (#12581)
pick 82e095ce95 DOC generative model description etc for LatentDirichletAllocation (#12216)
pick 62117f482e Bug with string dtype in OneHotEncoder with handle_unknown='ignore' (#12471)
pick 66507a25a4 MAINT Use CircleCI for linting (#12606)
pick 9ecada83d8 DOC fix TeX error in Davies Bouldin formula
pick 4ab60559c8 MNT Change deprecation for min_impurity_split from removal to changing the default (#12400)
pick d25da1be20 FIX make joblib utils private, and remove mentions of externals.joblib (#12345)
pick ac327c5ad6 [MRG] DOC Windows build dependencies (#12615)
pick 0b8650ab21 [MRG] set blockwise diagonals to zero for euclidean distance (#12612)
```
?

amueller committed:
FIX check_array dtype check for pandas series (#12625)

jnothman commented:
Thanks for looking that over, @amueller!

jnothman commented:
Added 5fd9e03 and b498ac7 and 94db3d9

> Shouldn't #12382 be included?

No. It applies to a 0.21 change to `count_nonzero`.

> drop 664ff34 ENH Prefer threads for IsolationForest (#12543)

I dropped this automatically because it modified v0.21.rst. Should we put it in?

jnothman commented:
> and then ... ?

Yes. Pushed

jnothman commented:
Can I document the release date as 23 Nov?

amueller commented:
aww after thanksgiving? ;)
I mean I was hoping for tomorrow but we don't really need to rush.

Maybe @ogrisel has an opinion on #12543?

adrinjalali commented:
I'll satisfy the the review points on #12626 in the morning => as far as the last blocker goes, should be done tomorrow!

jnothman commented:
I hadn't seen the progress on make_column_transformer. 21st Nov, then??

jnothman committed:
DOC whats new tweaks for 0.20.1

jnothman committed:
FIX Missing import

jnothman committed:
BLD PyPy needs joblib (#12629)

adrinjalali committed:
API Change the tuple order in make_column_transformer (#12626)

jnothman commented:
Renamed MRG. I leave the merge and release to @amueller, @ogrisel if that's alright.

rth commented:
Great that it's happening. For the person merging this, please don't forget not to squash when merging :)

qinhanmin2014 commented:
Hmm, restarted Circle CI, python 3 job is still failing, which means that we cannot generate the doc. The error message is different from master:
```
generating gallery for auto_examples/neural_networks... [ 25%] plot_mnist_filters.py
Killed
Makefile:43: recipe for target 'html' failed
make: *** [html] Error 137
Exited with code 2
```

rth commented:
> generating gallery for auto_examples/neural_networks... [ 25%] plot_mnist_filters.py
> Killed
> Makefile:43: recipe for target 'html' failed

Without any other output, does it mean a segfault? Restarting the Circle CI job with SSH access could help investigating ..

qinhanmin2014 commented:
> Without any other output, does it mean a segfault? Restarting the Circle CI job with SSH access could help investigating ..

Apologies I'm busy now and I think it's a blocker. @rth you also have the privilege.

rth commented:
No need to apologize, I was just thinking outloud. I won't be able to look into it before this evening either..

adrinjalali commented:
Of it also fails locally, I could give a hand.

qinhanmin2014 commented:
> Of it also fails locally, I could give a hand.

@adrinjalali It also fails locally? I though the failure only appears in Circle CI.

adrinjalali commented:
Sorry, typo, I meant "if" (autocorrect). I'm compiling the branch and building the docs locally now to see if I can replicate the issue.

lesteve commented:
I have launched a rebuild with SSH enabled. Let's see what happens.

qinhanmin2014 commented:
Things in my mind, feel free to pick or close
#12633 MNT Reomve LGTM wanings (apart from joblib) in 0.20.1
#12634 DOC Note that we have API changes in 0.20.1

jnothman commented:
It's very strange for this to crop up all of a sudden... Have you tried
clearing the conda instance from the cache?

It's hilarious that it's plot_stock_market again causing headaches...

rth commented:
Interesting problem. I suppose adding `nomkl` to CircleCI conda requirements might be a lazy fix, but it's more avoiding the problem than solving it.

Also FWIW alls tests pass on conda forge using this PR https://github.com/conda-forge/scikit-learn-feedstock/pull/80

lesteve commented:
> It's very strange for this to crop up all of a sudden... Have you tried clearing the conda instance from the cache? 

Not sure how I would do that to be honest. From a quick look at .circle/config.yml it does not look like we are caching the conda install but maybe I am missing something. What I tried:

```
conda clean -a -y # to be on the safe side
conda remove -n my-test --all -y
conda create -n my-test numpy ipython python=3.6 -y

SafetyError: The package for numpy-base located at /home/circleci/miniconda/pkgs/numpy-base-1.15.4-py36h81de0dd_0
appears to be corrupted. The path 'lib/python3.6/site-packages/numpy/linalg/linalg.py'
has a sha256 mismatch.
  reported sha256: e410474c4534f3b226ba7f16faab47331406e8841ed35e9827fbd61638785d91
  actual sha256: 84bf78e8b2ff62c946cd9e4934ffca1e5795f5bce93a4aaac60986cefcef0243
```

> It's hilarious that it's plot_stock_market again causing headaches...

Depends on your sense of humour I guess, but yeah I totally get your point ;-)

jnothman commented:
You hadn't mentioned the corrupted numpy-base before had you? It certainly looks suspicious!

lesteve commented:
> You hadn't mentioned the corrupted numpy-base before had you? It certainly looks suspicious!

Looking at it a bit more it may be a red herring not sure ... I was able to delete the cached packages (`conda clean -a` did not seem to work for some reason) and re-download them but that did not get rid of the segmentation fault.

I have tested that adding `nomkl` was fixing the numpy problem above so it may be a reasonable work-around indeed.

rth commented:
Overall it feels like a packaging issue upstream with numpy and MKL. Maybe open an issue at numpy with your minimal example from https://github.com/scikit-learn/scikit-learn/pull/12383#issuecomment-440612975 and use nomkl (which would also make builds faster)?

ogrisel commented:
> Maybe we can open an issue at numpy with your minimal example from #12383 (comment)

Has anyone tried and manage to reproduce the crash locally or is this just an issue on the circle ci host? The stock market example works with my numpy with openblas installed from pypi.

rth commented:
I can't reproduce the small example https://github.com/scikit-learn/scikit-learn/pull/12383#issuecomment-440618327 using the same docker image as in CircleCI. Have not tried to run the full build though, so versions might differ slightly..

lesteve commented:
> Overall it feels like a packaging issue upstream with numpy and MKL. Maybe open an issue at numpy with your minimal example from #12383 (comment)

The thing is I only managed to reproduce the error on the CircleCI machine ... I haven't tried reusing the same docker image to reproduce locally. I'll try first using nomkl in the CircleCI build.

qinhanmin2014 commented:
> I'll try first using nomkl in the CircleCI build.

A PR with ``[doc build]``?

amueller commented:
Such a great thing to wake up to lol. Why are there NaN in that array anyway?
@rth can you run the full build?
Looks like a caching issues indeed. I'll only be able to do it in an hour, but I would try to ssh into circle again and see if we can fix the numpy-core mismatch thing and/or clear the cache there.

Also: this branch is building right now, right? Where's the failure?

amueller commented:
For the record, the installed versions of numpy that are failing are
    numpy:                    1.15.4-py36h1d66e8a_0  
    numpy-base:               1.15.4-py36h81de0dd_0

rth commented:
I won't be able to now. Possibly some issue with the default conda channel cf  https://github.com/scikit-learn/scikit-learn/pull/12636#issuecomment-440668153 It also fails on master.

lesteve commented:
> Such a great thing to wake up to lol. Why are there NaN in that array anyway?

Good morning! Good question about the NaNs, not sure.

> Looks like a caching issues indeed. I'll only be able to do it in an hour, but I would try to ssh into circle again and see if we can fix the numpy-core mismatch thing and/or clear the cache there.

Which cache do you have in mind and how would you clear it? I would have thought that there were no cache for the miniconda install but maybe I am missing something.

amueller commented:
Looks like there is indeed no miniconda cache. Your ssh was down, so I restarted it and I'll try to have a look, but it'll be a bit slow.

amueller commented:
but doing ``nomkl`` does seem to fix it.

ogrisel commented:
> but doing nomkl does seem to fix it.

It did not fix it for the python 2 circle ci build in #12636 (first commit of that PR).

amueller commented:
Sorry edited: I didn't use conda-forge, I did ``conda install nomkl``.

amueller commented:
you said

> Also, if I do another rebase of master onto this branch, it is able to recognise which commits I've already transferred to the new branch

but that's not happening for me using your branch. Is that expected?

amueller commented:
ahh I think you mean  the rebase only shows the commits that are not included? that makes sense...

amueller commented:
honestly I have no idea and will just cherry-pick.

amueller commented:
It included some that you picked before, though. That was confusing to me.

qinhanmin2014 commented:
Hmm, so Circle CI fails in the PR and passes in 0.20.X? So many mysterious things.
